### ISSUE TITLE
BUG: datadog/notification-api Containers constantly exiting

### ISSUE THREAD

**Issue by ldraney** [department-of-veterans-affairs/notification-api#1699]
Value statement:
**As a** VA Notify Engineer 
**I want** to stop our containers from constantly restarting 
**So that** our system remains stable

**As a** Product Stakeholder
**I want** our service to be stable for our clients
**So that** we consistently provide the best service possible

## Description
Our Datadog and notificatino-api containers are constantly exiting out (see images).

### Additional Info and Resources

This may be to a task definition configuration or an issue with how the forwarder is configured, and I would suggest looking into container configuration and the latest image pushed to ECR first before looking at the forwarder configuration

## Engineering Checklist

- [x] Open support ticket with DD when ticket is picked up
- [x] Investigate Datadog image and task definitions; considering environment variables and ddtrace
- [x] Investigate when these exits started and if its related to the updated datadog image
- [x] Consider whether this is related to the forwarder

## Steps to Reproduce

This is an active problem that can be [seen here](https://vanotify.ddog-gov.com/containers?query=ecs_cluster_name%3Aprod-notification-cluster%20AND%20notification-api&overview-na-groups=false&selectedTopGraph=timeseries&to_ts=1711112308381&live=false):
<img width="1295" src="https://api.zenhub.com/attachedFiles/eyJfcmFpbHMiOnsibWVzc2FnZSI6IkJBaHBBMG9DQkE9PSIsImV4cCI6bnVsbCwicHVyIjoiYmxvYl9pZCJ9fQ==--6305ac281d7923f3a471b8fdf36ae5850c5a2b12/image%20%2814%29.png" alt="image (14).png" />
<img width="1510" src="https://api.zenhub.com/attachedFiles/eyJfcmFpbHMiOnsibWVzc2FnZSI6IkJBaHBBMFFDQkE9PSIsImV4cCI6bnVsbCwicHVyIjoiYmxvYl9pZCJ9fQ==--0cf3613e4b0c00b4eb8481a4c84ad22f33b851d4/image%20%2816%29.png" alt="image (16).png" />
<img width="907" src="https://api.zenhub.com/attachedFiles/eyJfcmFpbHMiOnsibWVzc2FnZSI6IkJBaHBBMGdDQkE9PSIsImV4cCI6bnVsbCwicHVyIjoiYmxvYl9pZCJ9fQ==--3950bd90a443a13bbeb8388e901275aedfad5343/image%20%2815%29.png" alt="image (15).png" />

## Impact/Urgency

This should be considered medium priority.  There seems to be no impact to our datadog currently, but because our containers are actively exiting and this is a critical tool for our SRE, this should be addresses sooner rather than later. 

## Expected Behavior

- [x] Datadog containers do not exit, and this is reflected on the containers page in Datadog.

**Comment by k-macmillan** [comment]
Investigated this and was unable to find any evidence the API containers were actually exiting. There are no logs showing gunicorn restarting. There are no logs in ECS I could find showing that the containers were restarting. The exit status that Datadog is reporting for API containers comes and goes.

The Datadog agent is 100% having issues though, and does need investigation. Possibly an upgrade to `ddtrace`.

**Comment by mjones-oddball** [comment]
Hey team! Please [add your planning poker estimate](https://app.zenhub.com/workspaces/VA-Notify-620d21369d810a00146ed9c8/issues/department-of-veterans-affairs/notification-api/1699?planning-poker) with Zenhub @cris-oddball @EvanParish @k-macmillan @kalbfled @ldraney @MackHalliday @mchlwellman

**Comment by npmartin-oddball** [comment]
@ldraney to reach out to DD, update ticket with response info...will inform next steps re: working ticket.

**Comment by ldraney** [comment]
Sent ddog a message, we'll see how they respond!  I'll then update this ticket comment and maybe additional info based on that response.

**Comment by ldraney** [comment]
Got a response from DDog team:

> Hi Lucas,
>  
> Can you confirm if your Agent deployment configuration has the following?
>  
> - Name: DD_PROCESS_AGENT_ENABLED
> Value: "true"
>  
> If the above is set, you will need to set the pidMode parameter to task in the task definition for us to properly execute according to our doc here.
>  
> Sincerely,
> Spencer Phillips | Solutions Engineer | Datadog | Business Hours: 10 a.m. to 10 p.m. ET, Monday to Friday

**Comment by ldraney** [comment]
From the AWS documentation for [PID mode](https://docs.aws.amazon.com/AmazonECS/latest/developerguide/task_definition_parameters.html#task_definition_pidmode):

> If task is specified, all containers within the specified task share the same process namespace.

**Comment by ldraney** [comment]
Which would need to be set up via the task definition here:
```
  "containerDefinitions": [
    {
      "name": "notification-api",
      "image": "{will-be-replaced-by-ci}",
	  "pidMode": "task",
```

**Comment by ldraney** [comment]
Okay, updated a branch and preparing to deploy.  Before I need to get a little setup in Datadog to verify the containers are still failing BEFORE/AFTER my deploy to dev

**Comment by ldraney** [comment]
I can't tell how Kyle got the "exited at" column in the above screenshot. 
Compared to what I see:
<img width="1280" src="https://api.zenhub.com/attachedFiles/eyJfcmFpbHMiOnsibWVzc2FnZSI6IkJBaHBBMFZDQkE9PSIsImV4cCI6bnVsbCwicHVyIjoiYmxvYl9pZCJ9fQ==--e51b8f430ebfd1bd75a9462e6e0b45c6556b2646/image.png" alt="image.png" />

**Comment by ldraney** [comment]
But the logs still clearly show exiting containers, which should be enough for now:
<img width="1278" src="https://api.zenhub.com/attachedFiles/eyJfcmFpbHMiOnsibWVzc2FnZSI6IkJBaHBBMGxDQkE9PSIsImV4cCI6bnVsbCwicHVyIjoiYmxvYl9pZCJ9fQ==--b59b990d3f353b4b6ee0cd99e2884786cf284e37/image.png" alt="image.png" />

**Comment by ldraney** [comment]
[deploying to dev](https://github.com/department-of-veterans-affairs/notification-api/actions/runs/8604876578)

**Comment by ldraney** [comment]
Had to fix task definition based on [AWS doc](https://docs.aws.amazon.com/AmazonECS/latest/developerguide/task-definition-template.html)

**Comment by ldraney** [comment]
The errors kept coming.  It may have been from the other tasks, so I will try pidmode on those as well. 

But first I want to make sure this is related to the DD_ PROCESS_AGENT_ ENABLED variable, as support suggested. 

I'm deleting that variable for now and redeploying to see if the errors stop. 

Then, I'll try PIDmode one more time.  Getting our system processes in fargate seems like a good idea, so I'll try to get that working while I'm here (not outside of scope - should have been the original functionality); but if it doesn't work we'll just consider bug fixed for now, I suppose.  Still thinking about it.

**Comment by ldraney** [comment]
[ Successful deploy between 1:27-1:29 MST](https://github.com/department-of-veterans-affairs/notification-api/actions/runs/8605501855) (celery finished deploying last), seeing outcome...

**Comment by ldraney** [comment]
The logs did stop!  Confirmed bug is related to DD_PROCESS_AGENT_ENABLED. 
<img width="1038" src="https://api.zenhub.com/attachedFiles/eyJfcmFpbHMiOnsibWVzc2FnZSI6IkJBaHBBNGhDQkE9PSIsImV4cCI6bnVsbCwicHVyIjoiYmxvYl9pZCJ9fQ==--22bec73ff171268b9aa078daa411ad29f7bfbd64/image.png" alt="image.png" />

***

Next step, try readding this with pidMode: task.  If not, this is an appropriate time to continue conversation with Datadog.

**Comment by ldraney** [comment]
[Successful deploy with pidmode: task](https://github.com/department-of-veterans-affairs/notification-api/actions/runs/8605640406)

and logs seem promising!  You can see errors have stopped for the last two deploys so far:

<img width="908" src="https://api.zenhub.com/attachedFiles/eyJfcmFpbHMiOnsibWVzc2FnZSI6IkJBaHBBNU5DQkE9PSIsImV4cCI6bnVsbCwicHVyIjoiYmxvYl9pZCJ9fQ==--9d61a8259d55ed078a48f66ab84c3ba1c09012cb/image.png" alt="image.png" />

I'm going to give it a little time, and also see if we are getting process information in Datadog as the [ddog documentation indicates we should](https://docs.datadoghq.com/integrations/ecs_fargate/?tab=webui#process-collection).

**Comment by ldraney** [comment]
We are indeed getting new process information:

### Before
<img width="579" src="https://api.zenhub.com/attachedFiles/eyJfcmFpbHMiOnsibWVzc2FnZSI6IkJBaHBBNkJDQkE9PSIsImV4cCI6bnVsbCwicHVyIjoiYmxvYl9pZCJ9fQ==--25f870ad2a66f4fa8df28db817fade05f84eb003/image.png" alt="image.png" />

### After
<img width="510" src="https://api.zenhub.com/attachedFiles/eyJfcmFpbHMiOnsibWVzc2FnZSI6IkJBaHBBNTFDQkE9PSIsImV4cCI6bnVsbCwicHVyIjoiYmxvYl9pZCJ9fQ==--8cbf4ed96abfa8a7d923a0755c234eec2ed22bed/image.png" alt="image.png" />

**Comment by ldraney** [comment]
Things are looking good!  I'm going to adjust the other environments, check on these pages again to confirm containers have stopped exiting, and open the PR with appropriate links in the "tests" section

**Comment by ldraney** [comment]
Having a conversation with Kyle related to the risks of this feature. 
Perhaps more notably, I noticed that the datadog container is marked as "essential", where it should actually be the notification-api container that is essential!  That should be addressed asap, so that datadog failures don't affect notification-api containers and tasks. 

https://docs.aws.amazon.com/AmazonECS/latest/developerguide/task_definition_parameters.html#ContainerDefinition-essential

**Comment by ldraney** [comment]
Kyle wants a more robust investigation to PidMode, Essential, and how we transfer PID to our application.

**Comment by ldraney** [comment]
[This page has been the most useful](https://aws.amazon.com/blogs/containers/announcing-additional-linux-controls-for-amazon-ecs-tasks-on-aws-fargate/)

A useful and relevant TLDR section is at the end, Caveats of sharing a process id namespace:
While a process id namespace can now be shared by containers in the same ECS task on AWS Fargate, there a few things to be aware of. We’ll walk through those caveats in the context of the application and sidecar ECS task defined previously:
A process in the sidecar container can observe, stop, or restart a process in the application container.
When sharing a process in an ECS task, a new pause process runs as PID 1 for the whole task.

So, that explains the pause task we see now. 
It also means that, as long as the datadog container has no functionality that can stop the notification-api container (I would be greatly surprised if it did), that notification-api should run smoothly, regardless of the status of the container. 
To address your point earlier that things have crashed due to datadog, this should be moreso attributed to our ddtrace setup, and I'm in communication with the datadog team on this.  This is because Fargate is essentially running docker-compose, each container has its own isolated resources, so even if one container crashed, it will not affect the other container unless there was a dependency relationship (e.g. notification-api depending upon datadog, which it doesn't). 

I'm also asking Datadog about their guidance on essential, as their own guide declares essential on both containers in the task, which doesn't make sense to me.

**Comment by ldraney** [comment]
From AWS customer support: 

> Setting pidMode to task would not cause all containers to fail by default. If, for example, you had a task with 3 containers with pidMode set to task where 2 of the tasks did something small like download a file and then exit. Having those containers exit would not automatically cause the whole task to fail even if they exited with a non-zero exit code.

**Comment by ldraney** [comment]
AWS support is asking for more information on how we use PID, though assumes we are using it for [graceful shutdown with ECS](https://aws.amazon.com/blogs/containers/graceful-shutdowns-with-ecs/).  Confirming this with Kyle.

**Comment by ldraney** [comment]
I see related PID settings in [gunicorn_config.py](https://github.com/department-of-veterans-affairs/notification-api/blob/master/gunicorn_config.py)

**Comment by ldraney** [comment]
Most useful documentation so far:
[Announcing additional Linux controls for Amazon ECS tasks on AWS Fargate](https://aws.amazon.com/blogs/containers/announcing-additional-linux-controls-for-amazon-ecs-tasks-on-aws-fargate/), which addresses some of our main concerns.

**Comment by ldraney** [comment]
Also, Datadog responded to me about essential: 
TLDR: we should have it on both containers, [as their guide suggests](https://www.datadoghq.com/blog/monitor-aws-fargate/#deploy-the-datadog-agent-on-aws-fargate): 

> Hi Lucas,
>  
> We mention specifying both containers as essential as the application container could still start without the Agent but this would cause the loss of monitoring. Marking the Agent as essential within the same task definition guarantees that the Agent is also started as a sidecar to the application. AWS mentions that at least one container should be marked as essential but this does not preclude additional containers within the task from being marked as such.
>  
> Sincerely,
> Spencer Phillips | Solutions Engineer | Datadog

**Comment by ldraney** [comment]
I continue to wait for AWS next response.  Meanwhile, I'm engaging in some DevOps study.

**Comment by ldraney** [comment]
I have been helping Mack for the last little bit. 
AWS support still hasn't responded.  Sent them a message:

> This is great!  We're awaiting your response.  I appreciate the details you can provide for our own documentation on this situation!

**Comment by ldraney** [comment]
[AWS response as of ten minutes ago](https://us-gov-west-1.console.amazonaws-us-gov.com/support/home#/case/?displayId=171267625700456&language=en)

**Comment by ldraney** [comment]
[successful deployment as of 2:25 MST](https://github.com/department-of-veterans-affairs/notification-api/actions/runs/8637289706).  [Checking logs](https://vanotify.ddog-gov.com/logs?query=%2Aexit%2A%20-host%3A%28%22%2Futility%2Fperf-staging-kafka-producer%22%20OR%20%22%2Futility%2Ftwistlock%22%20OR%20staging-notification-api-datadog-log-group%20OR%20perf-notification-api-datadog-log-group%20OR%20prod-notification-api-datadog-log-group%20OR%20%22%2Futility%2Fqa%22%29&agg_q=status%2Cservice&agg_q_source=base%2Cbase&cols=host%2Cservice&fromUser=true&index=%2A&messageDisplay=inline&refresh_mode=sliding&sort_m=%2C&sort_m_source=%2C&sort_t=%2C&storage=hot&stream_sort=desc&top_n=10%2C10&top_o=top%2Ctop&viz=pattern&x_missing=true%2Ctrue&from_ts=1712779718168&to_ts=1712780618168&live=true)

**Comment by ldraney** [comment]
[so far looking good](https://vanotify.ddog-gov.com/logs?query=%2Aexit%2A%20-%28host%3A%22%2Futility%2Fperf-staging-kafka-producer%22%20OR%20host%3A%22%2Futility%2Ftwistlock%22%20OR%20host%3Astaging-notification-api-datadog-log-group%20OR%20host%3Aperf-notification-api-datadog-log-group%20OR%20host%3Aprod-notification-api-datadog-log-group%20OR%20host%3A%22%2Futility%2Fqa%22%29%20service%3Acloudwatch%20status%3Ainfo%20%22process%20agent%20exited%20with%20code%202%20signal%200%20restarting%20in%202%20seconds%22&agg_q=status%2Cservice&agg_q_source=base%2Cbase&cols=host%2Cservice&fromUser=true&index=%2A&messageDisplay=inline&refresh_mode=sliding&sort_m=%2C&sort_m_source=%2C&sort_t=%2C&storage=hot&stream_sort=desc&top_n=10%2C10&top_o=top%2Ctop&viz=timeseries&x_missing=true%2Ctrue&from_ts=1712779718168&to_ts=1712780618168&live=true)

<img width="1164" src="https://api.zenhub.com/attachedFiles/eyJfcmFpbHMiOnsibWVzc2FnZSI6IkJBaHBBNnBPQkE9PSIsImV4cCI6bnVsbCwicHVyIjoiYmxvYl9pZCJ9fQ==--1fafec6ebea0f617739cff836a40dec1fbd72481/image.png" alt="image.png" />


Preparing for a sync with Kyle on this

**Comment by ldraney** [comment]
I'm going to remove essential from my PR and open this tomorrow morning for review

**Comment by ldraney** [comment]
Merged!  Moving this to the QA lane.

**Comment by ldraney** [comment]
I found the logs most helpful for development.  the above log search query with *pool* and *shutdown* proves the pidmode didn't interfere with graceful shutdown. 
or *exit* in the same dev-notification-api-log-group, with a timespan of a week, will show these errors coming and going as I messed with deployments, and ultimately gone away since the deploys today.  

Also, the containers are clearly not exiting anymore on [the containers page](https://vanotify.ddog-gov.com/containers?query=ecs_cluster_name%3Adev-notification-cluster&overview-na-groups=false&selectedTopGraph=timeseries&sort=container_memory%2CDESC); if the containers were exiting, this would be apparent in more recent start times for some failing containers, but since they are all the same since the last deployment, the fix appears to be working. : 

<img width="1114" src="https://api.zenhub.com/attachedFiles/eyJfcmFpbHMiOnsibWVzc2FnZSI6IkJBaHBBK1pTQkE9PSIsImV4cCI6bnVsbCwicHVyIjoiYmxvYl9pZCJ9fQ==--b18c660119c363e741ffaf314b0954033d86f1ec/image.png" alt="image.png" />
